File: Source/Engine/ReCrafted.Core/Core/Time.cpp
// ReCrafted (c) 2016-2019 Damian 'Erdroy' Korczowski. All rights reserved.

#include "Time.h"
#include "Common/Platform/Platform.h"

void Time::Initialize()
{
}

void Time::Shutdown()
{
}

void Time::OnFrame()
{
    const auto currentTime = Platform::GetMilliseconds();
    m_deltaTime = float((currentTime - m_lastUpdateTime) / 1000.0);
    m_time = float(currentTime / 1000.0);
    m_lastUpdateTime = currentTime;

    m_frames++;
}

float Time::CurrentTime()
{
    MAIN_THREAD_ONLY();
    return GetInstance()->m_time;
}

float Time::CurrentFixedTime()
{
    MAIN_THREAD_ONLY();
    return GetInstance()->m_fixedTime;
}

float Time::DeltaTime()
{
    MAIN_THREAD_ONLY();
    return GetInstance()->m_deltaTime;
}

float Time::FixedDeltaTime()
{
    MAIN_THREAD_ONLY();
    return GetInstance()->m_fixedDeltaTime;
}

uint32_t Time::CurrentFrame()
{
    MAIN_THREAD_ONLY();
    return GetInstance()->m_frames;
}

File: Source/Engine/ReCrafted.API/Core/Time.cs
// ReCrafted (c) 2016-2019 Damian 'Erdroy' Korczowski. All rights reserved.

namespace ReCrafted.API.Core
{
    /// <summary>
    ///     The time class. Provides engine's timing services.
    /// </summary>
    public static partial class Time
    {
        /// <summary>
        ///     Returns current frame engine's time.
        /// </summary>
        public static float CurrentTime => InternalCurrentTime();

        /// <summary>
        ///     Returns current frame fixed engine's time.
        /// </summary>
        public static float CurrentFixedTime => InternalCurrentFixedTime();

        /// <summary>
        ///     Returns the precise real time since startup in seconds.
        /// </summary>
        public static double Realtime => InternalRealtime();

        /// <summary>
        ///     Returns the time between last frame and current one.
        /// </summary>
        public static float DeltaTime => InternalDeltaTime();

        /// <summary>
        ///     Returns the fixed time. This time is being used as time step
        ///     for the FixedUpdate function calls.
        /// </summary>
        public static float FixedDeltaTime => InternalFixedDeltaTime();

        /// <summary>
        ///     Returns the frame alpha value.
        /// </summary>
        public static float FrameAlpha => InternalFrameAlpha();

        /// <summary>
        ///     The current frame number.
        /// </summary>
        public static uint Frame => InternalCurrentFrame();
    }
}
